sql/moz-fx-data-shared-prod/mozilla_vpn_derived/subscriptions_v1/query.sql

```diff
@@ -1,3 +1,147 @@
+CREATE TEMP FUNCTION udf_js_apple_receipt(receipt STRING)
+-- provide the full schema even though most of it is scrubbed
+RETURNS STRUCT<
+  environment STRING,
+  latest_receipt BYTES,
+  latest_receipt_info ARRAY<
+    STRUCT<
+      expires_date STRING,
+      expires_date_ms STRING,
+      expires_date_pst STRING,
+      in_app_ownership_type STRING,
+      is_in_intro_offer_period STRING,
+      is_trial_period STRING,
+      original_purchase_date STRING,
+      original_purchase_date_ms STRING,
+      original_purchase_date_pst STRING,
+      original_transaction_id STRING,
+      product_id STRING,
+      purchase_date STRING,
+      purchase_date_ms STRING,
+      purchase_date_pst STRING,
+      quantity STRING,
+      subscription_group_identifier STRING,
+      transaction_id STRING,
+      web_order_line_item_id STRING
+    >
+  >,
+  pending_renewal_info ARRAY<
+    STRUCT<
+      auto_renew_product_id STRING,
+      auto_renew_status STRING,
+      original_transaction_id STRING,
+      product_id STRING
+    >
+  >,
+  receipt STRUCT<
+    adam_id INT64,
+    app_item_id INT64,
+    application_version STRING,
+    bundle_id STRING,
+    download_id INT64,
+    in_app ARRAY<
+      STRUCT<
+        expires_date STRING,
+        expires_date_ms STRING,
+        expires_date_pst STRING,
+        in_app_ownership_type STRING,
+        is_in_intro_offer_period STRING,
+        is_trial_period STRING,
+        original_purchase_date STRING,
+        original_purchase_date_ms STRING,
+        original_purchase_date_pst STRING,
+        original_transaction_id STRING,
+        product_id STRING,
+        purchase_date STRING,
+        purchase_date_ms STRING,
+        purchase_date_pst STRING,
+        quantity STRING,
+        transaction_id STRING,
+        web_order_line_item_id STRING
+      >
+    >,
+    original_application_version STRING,
+    original_purchase_date STRING,
+    original_purchase_date_ms STRING,
+    original_purchase_date_pst STRING,
+    receipt_creation_date STRING,
+    receipt_creation_date_ms STRING,
+    receipt_creation_date_pst STRING,
+    receipt_type STRING,
+    request_date STRING,
+    request_date_ms STRING,
+    request_date_pst STRING,
+    version_external_identifier INT64,
+  >,
+  status INT64,
+> DETERMINISTIC
+LANGUAGE js
+AS
+  "return JSON.parse(receipt);";
+
+CREATE TEMP FUNCTION udf_scrub_apple_receipt(apple_receipt ANY TYPE) AS (
+  STRUCT(
+    apple_receipt.environment,
+    -- extract a list of start and end dates that doesn't overlap or repeat
+    ARRAY(
+      WITH exploded AS (
+        SELECT
+          date,
+          IF(
+            ANY_VALUE(date) OVER (
+              ORDER BY
+                UNIX_DATE(date) RANGE
+                BETWEEN 1 PRECEDING
+                AND 1 PRECEDING
+            ) IS NULL,
+            date,
+            NULL
+          ) AS start_date,
```

Comment:
This queries the start date by finding a value that doesn't have anything before it? Is this equivalent to `if(lag(date) over (order by date) is null, date, null)`?

Reply:
> This queries the start date by finding a value that doesn't have anything before it?

yeah. also note that this is using `RANGE`, not `ROWS`, so it's specifically checking if one day before `date` is present, not one row before.

> Is this equivalent to `if(lag(date) over (order by date) is null, date, null)`?

i'm not familiar with a `lag` aggregate function, but I could define one where that would be correct

Reply:
No need to redefine the function, I'm just trying to get a handle on how it worked. If LAG and LEAD do work in this context though, I think it would be a littler clearer in terms of intent.

Reply:
looking at the definitions, `LAG` and `LEAD` only do preceding row, not range